future-apps/payroll/contracts/Payroll.sol

```diff
@@ -0,0 +1,569 @@
+        // check that account doesn't exist
+        require(employeeIds[newAddress] == 0);
+        // check it's non-null address
+        require (newAddress != address(0));
+
+        var employee = employees[employeeId];
+        employeeIds[employee.accountAddress] = 0;
+        employee.accountAddress = newAddress;
+        employeeIds[newAddress] = employeeId;
+    }
+
+    /**
+     * @dev Change employee account address. To be called by Employee
+     * @notice Change employee account address
+     * @param newAddress New address to receive payments
+     */
+    function changeAddressByEmployee(address newAddress) public {
+        // check that account doesn't exist
+        require(employeeIds[newAddress] == 0);
+        // check it's non-null address
+        require (newAddress != address(0));
+        // check that employee exists (and matches)
+        var employeeId = employeeIds[msg.sender];
+        var employee = employees[employeeId];
+        // check that employee exists (and matches)
+        require(employee.accountAddress == msg.sender);
+
+        employee.accountAddress = newAddress;
+        employeeIds[newAddress] = employeeId;
+        employeeIds[msg.sender] = 0;
+    }
+
+    /* Aux functions */
+    /**
+     * @dev Get the Denomination exchange rate of a Token
+     * @notice Get the Denomination exchange rate of a Token
+     * @param token The token address
+     * @return denominationExchangeRate The exchange rate
+     */
+    function getExchangeRate(address token) public view returns(uint256 rate) {
+        rate = exchangeRates[token];
+    }
+
+    /**
+     * @dev Check that a token has the exchange rate already set
+     *      Internal function, needed to ensure that we have the rate before making a payment.
+     * @param token The token address
+     * @return True if we have the exchange rate, false otherwise
+     */
+    function checkExchangeRate(address token) internal view returns(bool) {
+        if (exchangeRates[token] == 0) {
+            return false;
+        }
+        return true;
+    }
+
+    function _addEmployee(
+        address accountAddress,
+        uint256 initialYearlyDenominationSalary,
+        string name,
+        uint256 startDate
+    )
+        internal
+    {
+        // check that account doesn't exist
+        require(employeeIds[accountAddress] == 0);
+
+        var employeeId = nextEmployee;
+        employees[employeeId] = Employee({
+            accountAddress: accountAddress,
+            denominationTokenSalary: initialYearlyDenominationSalary.toSecondDenominationToken(),
+            lastAllocation: 0,
+            lastPayroll: startDate,
+            name: name
+        });
+        // Ids mapping
+        employeeIds[accountAddress] = employeeId;
+        // update global variables
+        yearlyTotalPayroll = yearlyTotalPayroll.add(employees[employeeId].denominationTokenSalary.toYearlyDenomination());
```

Comment:
Do we need to make this a storage value? Seems like something that would be expensive to compute otherwise (loops involved) but do we use it in the logic of the contract anywhere or is it just 'called' to it?

Reply:
Yes, the storage was intended to avoid the loop. We use it for `calculatePayrollRunway` and `calculatePayrollBurnrate`.